pkg/sqlite/load_test.go

```diff
@@ -825,3 +820,211 @@ func TestAddBundlePropertiesFromAnnotations(t *testing.T) {
+							{
+								Name:           "a",
+								CurrentCSVName: "csv-a",
+							},
+							{
+								Name:           "b",
+								CurrentCSVName: "csv-a",
+							},
+						},
+						DefaultChannelName: "a",
+					},
+					{
+						PackageName: "pkg-1",
+						Channels: []registry.PackageChannel{
+							{
+								Name:           "a",
+								CurrentCSVName: "csv-b",
+							},
+							{
+								Name:           "b",
+								CurrentCSVName: "csv-b",
+							},
+						},
+						DefaultChannelName: "a",
+					},
+				},
+			},
+			args: args{
+				bundle: "csv-a",
+				pkg:    "pkg-0",
+			},
+			expected: expected{
+				bundles: map[string]struct{}{
+					"pkg-1/a/csv-b": {},
+					"pkg-1/b/csv-b": {},
+				},
+			},
+		},
+		{
+			description: "ChannelHead/WithReplacement",
+			fields: fields{
+				bundles: []*registry.Bundle{
+					newBundle(t, "csv-a", "pkg-0", []string{"a", "b"}, newUnstructuredCSV(t, "csv-a", "")),
+					newBundle(t, "csv-aa", "pkg-0", []string{"b"}, newUnstructuredCSV(t, "csv-aa", "csv-a")),
+				},
+				pkgs: []registry.PackageManifest{
+					{
+						PackageName: "pkg-0",
+						Channels: []registry.PackageChannel{
+							{
+								Name:           "a",
+								CurrentCSVName: "csv-a",
+							},
+							{
+								Name:           "b",
+								CurrentCSVName: "csv-aa",
+							},
+						},
+						DefaultChannelName: "b",
+					},
+				},
+			},
+			args: args{
+				bundle: "csv-a",
+				pkg:    "pkg-0",
+			},
+			expected: expected{
+				err: fmt.Errorf("cannot overwrite bundle csv-a from package pkg-0: replaced by csv-aa on channel b"),
+				bundles: map[string]struct{}{
+					"pkg-0/a/csv-a":  {},
+					"pkg-0/b/csv-a":  {},
+					"pkg-0/b/csv-aa": {},
+				},
+			},
+		},
+		{
+			description: "ChannelHead",
+			fields: fields{
+				bundles: []*registry.Bundle{
+					newBundle(t, "csv-a", "pkg-0", []string{"a", "b"}, newUnstructuredCSVWithSkips(t, "csv-a", "csv-b", "csv-c")),
+					newBundle(t, "csv-b", "pkg-0", []string{"b", "d"}, newUnstructuredCSV(t, "csv-b", "")),
+					newBundle(t, "csv-d", "pkg-0", []string{"d"}, newUnstructuredCSV(t, "csv-d", "csv-b")),
+					newBundle(t, "csv-c", "pkg-0", []string{"c"}, newUnstructuredCSV(t, "csv-c", "")),
+				},
+				pkgs: []registry.PackageManifest{
+					{
+						PackageName: "pkg-0",
+						Channels: []registry.PackageChannel{
+							{
+								Name:           "a",
+								CurrentCSVName: "csv-a",
+							},
+							{
+								Name:           "b",
+								CurrentCSVName: "csv-a",
+							},
+							{
+								Name:           "c",
+								CurrentCSVName: "csv-c",
+							},
+							{
+								Name:           "d",
+								CurrentCSVName: "csv-d",
+							},
+						},
+						DefaultChannelName: "a",
+					},
+				},
+			},
+			args: args{
+				bundle: "csv-a",
+				pkg:    "pkg-0",
+			},
+			expected: expected{
+				err: nil,
+				bundles: map[string]struct{}{
+					"pkg-0/a/csv-b": {},
+					"pkg-0/b/csv-b": {},
+					"pkg-0/a/csv-c": {},
```

Comment:
So `csv-b` and `csv-c` end up in channel A even after bundle `a` is removed. In this state, there would be an error because `csv-b` and `csv-c` are both in channel `a`, but they're in different replaces chains.

Is the assumption that the new overwrite bundle will resolve this situation? Or do we just rebuild the graph from scratch based on highest semver, and this intermediate state just doesn't matter?

Reply:
This is an issue when you have a new bundle overwriting an existing one in a new set of channels but not keeping the same edges as a subset of its new edges. The channel head selection heuristic would drop whichever one of b/c was of a lower semver from the a/b channels, but since we're checking for dropped bundles, we'd complain on either case.

So we'd be adding another hidden requirement to overwrites: new bundles must have edges to their old bundles on at least the channels they previously had those edges on.

If overwritten bundles are mainly for modifying bundle data rather than the graph metadata, this is a non-issue. If they do modify graph metadata, we'd at least warn when bundles get dropped.